config/v1alpha2/types_insights.go

```diff
@@ -0,0 +1,115 @@
+	// +optional
+	Status InsightsDataGatherStatus `json:"status"`
+}
+
+type InsightsDataGatherSpec struct {
+	// gatherConfig spec attribute includes all the configuration options related to
+	// gathering of the Insights data and its uploading to the ingress.
+	// +optional
+	GatherConfig GatherConfig `json:"gatherConfig,omitempty"`
+}
+
+type InsightsDataGatherStatus struct{}
+
+// gatherConfig provides data gathering configuration options.
+type GatherConfig struct {
+	// dataPolicy allows user to enable additional global obfuscation of the IP addresses and base domain
+	// in the Insights archive data. Valid values are "ClearText" and "ObfuscateNetworking".
+	// When set to ClearText the data is not obfuscated.
+	// When set to ObfuscateNetworking the IP addresses and the cluster domain name are obfuscated.
+	// When omitted, this means no opinion and the platform is left to choose a reasonable default, which is subject to change over time.
+	// The current default is ClearText.
+	// +optional
+	DataPolicy DataPolicy `json:"dataPolicy,omitempty"`
+	// gatherers is a list of gatherers configurations.
+	// The particular gatherers IDs can be found at https://github.com/openshift/insights-operator/blob/master/docs/gathered-data.md.
+	// Run the following command to get the names of last active gatherers:
+	// "oc get insightsoperators.operator.openshift.io cluster -o json | jq '.status.gatherStatus.gatherers[].name'"
+	// +kubebuilder:validation:MaxItems=100
+	// +listType=atomic
+	// +optional
+	Gatherers []GathererConfig `json:"gatherers,omitempty"`
+}
+
+// state declares valid gatherer state types.
+// +kubebuilder:validation:Enum="";Enabled;Disabled
+type GathererState string
+
+// gathererConfig allows to configure specific gatherers
+type GathererConfig struct {
+	// name is the name of specific gatherer
+	// +kubebuilder:validation:MaxLength=256
+	// +required
+	Name string `json:"name"`
+	// state allows you to configure specific gatherer. Valid values are "Enabled", "Disabled" and omitted.
+	// When omitted, this means no opinion and the platform is left to choose a reasonable default.
+	// The current default is Enabled.
+	// +optional
+	State GathererState `json:"state"`
+}
+
+const (
+	// No data obfuscation
+	ClearText DataPolicyOption = "ClearText"
+	// IP addresses and cluster domain name are obfuscated
+	ObfuscateNetworking DataPolicyOption = "ObfuscateNetworking"
+	// Gatherer state marked as disabled, which means that the gatherer will not run.
+	Disabled GathererState = "Disabled"
+	// Gatherer state marked as enabled, which means that the gatherer will run.
+	Enabled GathererState = "Enabled"
+)
+
+// dataPolicyOption declares valid data policy options
+// +kubebuilder:validation:Enum="";ClearText;ObfuscateNetworking
+type DataPolicyOption string
+
+// DataPolicy is a list of data policy options
+// +listType=set
+// +kubebuilder:validation:MaxItems=5
```

Comment:
There are only two valid options for this value, so, max is 2?

We are typically avoiding the `listType` `set` rule in favour of using CEL with an atomic list, we've run into issues before

Reply:
Can you provide more information for the usage of CEL and atomic usage instead of `listType=set`? The DataPolicy should be a simple list of strings. However when I am thinking about it we will probably need to make sure that there are not both `ClearText` and `ObfuscateNetworking` at the same time defined -> `["ClearText", "ObfuscateNetworking"]`. We made it into a list because there is one more option right now `WorkloadNames`  and to keep it in sync with our ConfigMap. (I had forgotten to add the option here...)

Reply:
To learn about CEL, take a look at https://kubernetes.io/docs/reference/using-api/cel/

Specifically here, Set has some issues with lists of objects (I know this is list of string), so we are staying away from it in general and using CEL to validate the uniqueness, you can add a rule like

```
self.exists_one(x, self.exists(y, x == y))
```

Which ensures uniqueness of objects within the list